File: api/routes/sessions.py
"""API routes for chat session management."""
import uuid
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, delete, func

from core.database import get_db
from core.logging import get_logger
from core.utils import to_iso8601
from models.chat import CHAT_TITLE_PLACEHOLDER, ChatSession, ChatMessage, ChatUsage
from routes.schemas import (
    ChatSessionResponse,
    CreateSessionRequest,
    ChatMessageResponse,
    PaginatedMessagesResponse,
    ChatContextResponse,
    DeleteSessionResponse,
    DeleteAllSessionsResponse,
)

router = APIRouter()
logger = get_logger(__name__)


@router.get("/", response_model=list[ChatSessionResponse])
async def list_sessions(
    db: AsyncSession = Depends(get_db)
):
    """Get list of all chat sessions, ordered by most recently updated"""
    logger.info("Fetching all chat sessions")
    try:
        result = await db.execute(
            select(ChatSession).order_by(ChatSession.updated_at.desc())
        )
        sessions = result.scalars().all()
        
        return [ChatSessionResponse(
            id=str(session.id),
            title=session.title,
            agent_id=session.agent_id or "hello-world",
            agent_name=session.agent_name,
            created_at=to_iso8601(session.created_at),
            updated_at=to_iso8601(session.updated_at),
            sdk_session_id=session.sdk_session_id,
        ) for session in sessions]
    except Exception as e:
        logger.error(f"Error fetching sessions: {e}", exc_info=True)
        raise HTTPException(status_code=500, detail=f"Error fetching sessions: {str(e)}")

@router.post("/", response_model=ChatSessionResponse)
async def create_session(
    request: CreateSessionRequest,
    db: AsyncSession = Depends(get_db)
):
    """Create a new chat session with the specified agent"""
    logger.info(f"Creating new session for agent {request.agent_id}")
    try:
        session = ChatSession(
            id=str(uuid.uuid4()),
            title=request.title or CHAT_TITLE_PLACEHOLDER,
            agent_id=request.agent_id,
            agent_name=request.agent_name,
        )
        db.add(session)
        await db.commit()
        await db.refresh(session)
        
        logger.info(f"Created session {session.id} for agent {request.agent_id}")
        
        return ChatSessionResponse(
            id=str(session.id),
            title=session.title,
            agent_id=session.agent_id,
            agent_name=session.agent_name,
            created_at=to_iso8601(session.created_at),
            updated_at=to_iso8601(session.updated_at),
            sdk_session_id=session.sdk_session_id,
        )
    except Exception as e:
        logger.error(f"Error creating session: {e}", exc_info=True)
        await db.rollback()
        raise HTTPException(status_code=500, detail=f"Error creating session: {str(e)}")

@router.get("/{chat_id}/messages", response_model=PaginatedMessagesResponse)
async def get_chat_messages(
    chat_id: str,
    db: AsyncSession = Depends(get_db),
    limit: int = Query(30, ge=1, le=100, description="Number of messages to return"),
    offset: int = Query(0, ge=0, description="Number of messages to skip")
):
    """Get paginated messages for a specific chat session
    
    Returns messages ordered by created_at ASC (oldest first).
    Default behavior: returns last 30 messages when offset=0.
    For infinite scroll: calculate offset based on total count minus desired messages.
    """
    logger.info(f"Fetching messages for chat {chat_id} with limit={limit}, offset={offset}")
    try:
        # Verify session exists
        session_result = await db.execute(
            select(ChatSession).where(ChatSession.id == chat_id)
        )
        session = session_result.scalar_one_or_none()
        if not session:
            raise HTTPException(status_code=404, detail=f"Chat session {chat_id} not found")
        
        # Get total count of messages
        count_result = await db.execute(
            select(func.count(ChatMessage.id))
            .where(ChatMessage.chat_id == chat_id)
        )
        total = count_result.scalar() or 0
        
        # If no messages, return empty response
        if total == 0:
            return PaginatedMessagesResponse(
                messages=[],
                total=0,
                limit=limit,
                offset=0,
                has_more=False
            )
        
        # Calculate actual offset for "last N messages" behavior
        # When offset=0, we want the last `limit` messages
        # We'll order by DESC, take the last N, then reverse to ASC order
        if offset == 0:
            # Get the last `limit` messages (newest first)
            result = await db.execute(
                select(ChatMessage)
                .where(ChatMessage.chat_id == chat_id)
                .order_by(ChatMessage.created_at.desc())
                .limit(limit)
            )
            messages = list(result.scalars().all())
            # Reverse to get oldest first (for display)
            messages.reverse()
            actual_offset = max(0, total - limit)
        else:
            # For pagination (loading older messages), fetch from offset
            result = await db.execute(
                select(ChatMessage)
                .where(ChatMessage.chat_id == chat_id)
                .order_by(ChatMessage.created_at.asc())
                .offset(offset)
                .limit(limit)
            )
            messages = result.scalars().all()
            actual_offset = offset
        
        message_responses = [ChatMessageResponse(
            id=str(msg.id),
            chat_id=str(msg.chat_id),
            role=msg.role,
            content=msg.content,
            created_at=to_iso8601(msg.created_at),
            metadata=getattr(msg, 'message_metadata', None)  # Map message_metadata to metadata in API response
        ) for msg in messages]
        
        has_more = actual_offset + len(messages) < total
        
        return PaginatedMessagesResponse(
            messages=message_responses,
            total=total,
            limit=limit,
            offset=actual_offset,
            has_more=has_more
        )
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error fetching messages: {e}", exc_info=True)
        raise HTTPException(status_code=500, detail=f"Error fetching messages: {str(e)}")

@router.get("/{chat_id}/context", response_model=ChatContextResponse)
async def get_chat_context(
    chat_id: str,
    db: AsyncSession = Depends(get_db)
):
    """Get aggregated usage statistics (context) for a chat session"""
    logger.info(f"Fetching context for chat {chat_id}")
    try:
        # Verify session exists
        session_result = await db.execute(
            select(ChatSession).where(ChatSession.id == chat_id)
        )
        session = session_result.scalar_one_or_none()
        if not session:
            raise HTTPException(status_code=404, detail=f"Chat session {chat_id} not found")
        
        # Aggregate usage statistics
        result = await db.execute(
            select(ChatUsage)
            .where(ChatUsage.chat_id == chat_id)
        )
        usage_records = result.scalars().all()
        
        total_input = sum(u.input_tokens for u in usage_records)
        total_output = sum(u.output_tokens for u in usage_records)
        total_tokens = sum(u.total_tokens for u in usage_records)
        total_cost = sum(u.cost_usd for u in usage_records)

        return ChatContextResponse(
            chat_id=chat_id,
            total_input_tokens=total_input,
            total_output_tokens=total_output,
            total_tokens=total_tokens,
            total_cost_usd=total_cost,
            agent_id=session.agent_id or "hello-world",
            agent_name=session.agent_name,
            context_window=200000  # Claude default context window
        )
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error fetching context: {e}", exc_info=True)
        raise HTTPException(status_code=500, detail=f"Error fetching context: {str(e)}")

@router.delete("/{chat_id}", response_model=DeleteSessionResponse)
async def delete_session(
    chat_id: str,
    db: AsyncSession = Depends(get_db)
):
    """Delete a specific chat session and all its messages (cascade delete)"""
    logger.info(f"Deleting chat session {chat_id}")
    try:
        # Verify session exists
        result = await db.execute(
            select(ChatSession).where(ChatSession.id == chat_id)
        )
        session = result.scalar_one_or_none()
        
        if not session:
            raise HTTPException(status_code=404, detail=f"Chat session {chat_id} not found")
        
        # Delete using delete statement (cascade will handle related records)
        await db.execute(delete(ChatSession).where(ChatSession.id == chat_id))
        await db.commit()
        
        logger.info(f"Successfully deleted chat session {chat_id}")
        return DeleteSessionResponse(
            message=f"Chat session {chat_id} deleted successfully",
            deleted_id=chat_id
        )
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error deleting session: {e}", exc_info=True)
        await db.rollback()
        raise HTTPException(status_code=500, detail=f"Error deleting session: {str(e)}")

@router.delete("/", response_model=DeleteAllSessionsResponse)
async def delete_all_sessions(
    db: AsyncSession = Depends(get_db)
):
    """Delete all chat sessions and their associated messages"""
    logger.info("Deleting all chat sessions")
    try:
        # Get count before deletion
        count_result = await db.execute(select(func.count(ChatSession.id)))
        deleted_count = count_result.scalar() or 0
        
        # Delete all sessions (cascade will handle related records)
        await db.execute(delete(ChatSession))
        await db.commit()
        
        logger.info(f"Successfully deleted {deleted_count} chat session(s)")
        return DeleteAllSessionsResponse(
            message="All chat sessions deleted successfully",
            deleted_count=deleted_count
        )
    except Exception as e:
        logger.error(f"Error deleting all sessions: {e}", exc_info=True)
        await db.rollback()
        raise HTTPException(status_code=500, detail=f"Error deleting all sessions: {str(e)}")

File: frontend/app/(chat)/actions/sessions.ts
"use server";

import { z } from "zod";
import { revalidatePath } from "next/cache";
import { backend, BackendError } from "@/lib/api/backend";
import {
  ChatSessionResponseSchema,
  PaginatedMessagesResponseSchema,
  ChatContextResponseSchema,
  DeleteSessionResponseSchema,
  DeleteAllSessionsResponseSchema,
  type ChatSessionResponse,
  type CreateSessionRequest,
  type PaginatedMessagesResponse,
  type ChatContextResponse,
  type DeleteSessionResponse,
  type DeleteAllSessionsResponse,
} from "@/lib/api/contracts";
import type { ActionResult } from "./types";

// ============================================================================
// Server Actions
// ============================================================================

/**
 * Get all chat sessions, ordered by most recently updated
 * Backend: GET /api/v1/sessions/
 */
export async function getSessions(): Promise<ActionResult<ChatSessionResponse[]>> {
  try {
    const data = await backend.get<ChatSessionResponse[]>("/sessions/");
    const validatedData = z.array(ChatSessionResponseSchema).parse(data);
    return { success: true, data: validatedData };
  } catch (error) {
    console.error("[getSessions] Error:", error);

    if (error instanceof BackendError) {
      return {
        success: false,
        error: error.message,
        code: error.statusCode,
      };
    }

    if (error instanceof z.ZodError) {
      console.error("[getSessions] Validation error:", error.errors);
      return { success: false, error: "Invalid response from server" };
    }

    return { success: false, error: "Failed to fetch sessions" };
  }
}

/**
 * Create a new chat session with the specified agent
 * Backend: POST /api/v1/sessions/
 */
export async function createSession(
  agentId: string,
  agentName: string,
  title?: string
): Promise<ActionResult<ChatSessionResponse>> {
  try {
    const requestBody: CreateSessionRequest = {
      agent_id: agentId,
      agent_name: agentName,
      title,
    };
    const data = await backend.post<ChatSessionResponse>("/sessions/", requestBody);
    const validatedData = ChatSessionResponseSchema.parse(data);

    // Revalidate paths that depend on sessions data
    revalidatePath("/");
    revalidatePath(`/chat/${validatedData.id}`);

    return { success: true, data: validatedData };
  } catch (error) {
    console.error("[createSession] Error:", error);

    if (error instanceof BackendError) {
      return {
        success: false,
        error: error.message,
        code: error.statusCode,
      };
    }

    if (error instanceof z.ZodError) {
      console.error("[createSession] Validation error:", error.errors);
      return { success: false, error: "Invalid response from server" };
    }

    return { success: false, error: "Failed to create session" };
  }
}

/**
 * Get paginated messages for a chat session
 * Backend: GET /api/v1/sessions/{chat_id}/messages
 */
export async function getMessages(
  chatId: string,
  limit = 30,
  offset = 0
): Promise<ActionResult<PaginatedMessagesResponse>> {
  try {
    const url = `/sessions/${chatId}/messages?limit=${limit}&offset=${offset}`;
    const data = await backend.get<PaginatedMessagesResponse>(url);
    const validatedData = PaginatedMessagesResponseSchema.parse(data);
    return { success: true, data: validatedData };
  } catch (error) {
    console.error(`[getMessages] Error for chat ${chatId}:`, error);

    if (error instanceof BackendError) {
      // For 404 (session not found), return error
      if (error.isNotFound()) {
        return {
          success: false,
          error: "Chat session not found",
          code: 404,
        };
      }

      return {
        success: false,
        error: error.message,
        code: error.statusCode,
      };
    }

    if (error instanceof z.ZodError) {
      console.error("[getMessages] Validation error:", error.errors);
      return { success: false, error: "Invalid response from server" };
    }

    return { success: false, error: "Failed to fetch messages" };
  }
}

/**
 * Get usage context/statistics for a chat session
 * Backend: GET /api/v1/sessions/{chat_id}/context
 */
export async function getChatContext(
  chatId: string
): Promise<ActionResult<ChatContextResponse>> {
  try {
    const data = await backend.get<ChatContextResponse>(`/sessions/${chatId}/context`);
    const validatedData = ChatContextResponseSchema.parse(data);
    return { success: true, data: validatedData };
  } catch (error) {
    console.error(`[getChatContext] Error for chat ${chatId}:`, error);

    if (error instanceof BackendError) {
      // For 404, return error so chat page can redirect
      if (error.isNotFound()) {
        return {
          success: false,
          error: "Chat session not found",
          code: 404,
        };
      }

      return {
        success: false,
        error: error.message,
        code: error.statusCode,
      };
    }

    if (error instanceof z.ZodError) {
      console.error("[getChatContext] Validation error:", error.errors);
      return { success: false, error: "Invalid response from server" };
    }

    return { success: false, error: "Failed to fetch chat context" };
  }
}

/**
 * Delete a specific chat session
 * Backend: DELETE /api/v1/sessions/{chat_id}
 */
export async function deleteSession(
  chatId: string
): Promise<ActionResult<DeleteSessionResponse>> {
  try {
    const data = await backend.delete<DeleteSessionResponse>(`/sessions/${chatId}`);
    const validatedData = DeleteSessionResponseSchema.parse(data);
    return { success: true, data: validatedData };
  } catch (error) {
    console.error(`[deleteSession] Error for chat ${chatId}:`, error);

    if (error instanceof BackendError) {
      return {
        success: false,
        error: error.message,
        code: error.statusCode,
      };
    }

    if (error instanceof z.ZodError) {
      console.error("[deleteSession] Validation error:", error.errors);
      return { success: false, error: "Invalid response from server" };
    }

    return { success: false, error: "Failed to delete session" };
  }
}

/**
 * Delete all chat sessions
 * Backend: DELETE /api/v1/sessions/
 */
export async function deleteAllSessions(): Promise<ActionResult<DeleteAllSessionsResponse>> {
  try {
    const data = await backend.delete<DeleteAllSessionsResponse>("/sessions/");
    const validatedData = DeleteAllSessionsResponseSchema.parse(data);
    return { success: true, data: validatedData };
  } catch (error) {
    console.error("[deleteAllSessions] Error:", error);

    if (error instanceof BackendError) {
      return {
        success: false,
        error: error.message,
        code: error.statusCode,
      };
    }

    if (error instanceof z.ZodError) {
      console.error("[deleteAllSessions] Validation error:", error.errors);
      return { success: false, error: "Invalid response from server" };
    }

    return { success: false, error: "Failed to delete all sessions" };
  }
}
